expo/eas-cli
# [eas-cli] Initialize analytics consistently and move analytics into context

# Checklist


# Why

The analytics system has been very fragile in the past. Things like always logging with a random uuid have plagued us numerous times.

This PR makes the analytics system non-global and instead injects it, which allows for consistent initialization and identification of the user to rudderstack/amplitude.

# How

This does a few things:
1. Make AnalyticsManager an interface, and only expose that interface to the commands. The internals of the manager system will switch between a no-op logger and the real one based on the user's analytics opt-out preferences.
2. Instantiate that interface when the command runs and inject it into contexts and commands. This way we can guarantee that it is initialized and the user identification has already occurred.
3. Only expose that interface to commands via a context for them to ensure that it is initialized as well.

# Test Plan

1. Run tests (they cover some of this).

# Manual test

Manually test to ensure that the user attribution is correct in all scenarios:

1. `export EXPO_STAGING=1`
2. Turn on live events in https://analytics.amplitude.com/expo/activity (with `Unified - dev` project selected)
3. While logged-in, run some commands, see that my user is getting `action` events.
4. Log out.
5. Run a command that requires login, see that once I log in the `action` event is correctly attributed to my existing amplitude user.
6. Log out.
7. Delete my `~/.expo-staging/state.json` and `~/Library/Preferences/eas-cli-nodejs/user-settings.json` (preference files) `analyticsDeviceId` key (remove persistent device IDs)
8. Run a command that requires login, see that once I log in the `action` event is correctly attributed to my existing amplitude user even though I have a new persistent device ID.
9. Log out
10. Repeat step 7 (delete persistent IDs)
11. Run a command that doesn't require login (`diagnostics`), see in amplitude that a new user is created.
12. Run a command that does require login, see that new user that was created is merged into my old user.
